samples/file-encryptor/enclave/openssl_src/keys.cpp

```diff
@@ -0,0 +1,438 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include <string.h>
+
+#include "aes_context.h" // Must be included before common/encryptor.h
+#include "common/encryptor.h"
+#include "common/trace.h"
+
+#define ENCRYPT_OPERATION true
+#define DECRYPT_OPERATION false
+
+void ecall_dispatcher::dump_data(
+    const char* name,
+    unsigned char* data,
+    size_t data_size)
+{
+    TRACE_ENCLAVE("Data name: %s", name);
+    for (size_t i = 0; i < data_size; i++)
+    {
+        TRACE_ENCLAVE("[%ld]-0x%02X", i, data[i]);
+    }
+    TRACE_ENCLAVE("\n");
+}
+
+// Compute the sha256 hash of given data.
+int ecall_dispatcher::Sha256(
+    const uint8_t* data,
+    size_t data_size,
+    uint8_t sha256[32])
+{
+    int ret = 1;
+    unsigned int sha_length_in_bytes = 32;
+    EVP_MD_CTX* mdctx;
+
+    if ((mdctx = EVP_MD_CTX_new()) == NULL)
+    {
+        TRACE_ENCLAVE("EVP_MD_CTX_new failed");
+        goto exit;
+    }
+
+    if (!(ret = EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL)))
+    {
+        TRACE_ENCLAVE(
+            "EVP_DigestInit_ex SHA-256 failed with returncode %d", ret);
+        goto exit;
+    }
+
+    if (!(ret = EVP_DigestUpdate(mdctx, data, data_size)))
+    {
+        TRACE_ENCLAVE("EVP_DigestUpdate failed with returncode %d", ret);
+        goto exit;
+    }
+
+    if (!(ret = EVP_DigestFinal_ex(mdctx, sha256, &sha_length_in_bytes)))
+    {
+        TRACE_ENCLAVE("EVP_DigestFinal_ex failed with return code %d", ret);
+        goto exit;
+    }
+
+    ret = 0;
+exit:
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+// This routine uses the openssl library to derive an AES key from the input
+// password and produce a password based key.
+int ecall_dispatcher::generate_password_key(
+    const char* password,
+    unsigned char* salt,
+    unsigned char* key,
+    unsigned int key_len)
+{
+    int ret = 1;
+
+    if (PKCS5_PBKDF2_HMAC(
+            (const char*)password,         // password
+            strlen((const char*)password), // password length
```

Comment:
strlen returns a size_t, but https://www.openssl.org/docs/man1.1.1/man3/PKCS5_PBKDF2_HMAC.html says this takes an int, so you need a cast to avoid compiler warnings

Reply:
implemented.